The code was previously building a new LocationLevel for each level v…

rma-rripken opened the pull request:
…alue.

I was looking into the NAB reported issue with East Sidney.Elev.Inst.0.Top of Conservation and I got CDA to generate a buggy json response.  
[response_1710176111087.json](https://github.com/USACE/cwms-data-api/files/14562431/response_1710176111087.json)

I duplicates the level for each level value in a nasty way.

I think these changes will fix it but I don't have a dataset handy to sufficiently test it locally.

rma-rripken committed:
Added ability to build urls to value instead of embedding large values.
Read from the ctx.body as an InputStream.
Added an png example.
Added url fields
Added methods to get requiredParams and required ZonedDateTimes.  These methods throw an exception if the required param isn't found.  Added exception block to ApiServlet that lets the users know what parameter wasn't found.
Added some tests.
Switching to UnaryOperator and Predicate<ResultSet>.
Including url-template for clobs
Passing contextPath to TextTimeSeriesController in-order to allow controller to build up url to clobs.
Util for doing simple {name} string replacement

rma-rripken committed:
The code was previously building a new LocationLevel for each level value.

gei-zchapman committed:
Implemented forecast DTOs

- Fields, constructors, and getters
- Round trip unit tests
- File parsing unit tests

rma-rripken commented:
https://github.com/USACE/cwms-data-api/issues/586

rma-rripken committed:
Merge remote-tracking branch 'origin/develop' into bugfix/level_duplicates

rma-rripken committed:
The dao currently only supports SI.  If EN is passed to the dao the units currently come back as SI units, better to throw an exception until it gets fixed

rma-rripken committed:
The dao currently only supports SI.  If EN is passed to the dao the units currently come back as SI units, better to throw an exception until it gets fixed

rma-rripken committed:
checkstyle fixes

rma-rripken committed:
checkstyle fixes

rma-rripken committed:
Moving the seasonal values to the end of the json.

rma-rripken committed:
Refactored code to make flow clearer.  Selecting from only the needed columns.  Added CALENDAR_OFFSET to the orderBy clause so that the seasonal values would be returned in order.  The Condition changes from the previous commit were incorrect. Because of how the JDom classes work the dao currently only supports SI

rma-rripken committed:
Refactored code to make flow clearer.  Selecting from only the needed columns.  Added CALENDAR_OFFSET to the orderBy clause so that the seasonal values would be returned in order.  The Condition changes from the previous commit were incorrect. Because of how the JDom classes work the dao currently only supports SI

rma-rripken committed:
checkstyle fixes. moved overloaded methods next to each other.

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/api/LevelsController.java:
```diff
@@ -268,6 +298,14 @@ public void getAll(Context ctx) {
+                    unit = UnitSystem.SI.getValue();
+                }
+                if(!UnitSystem.SI.getValue().equals(unit)) {
+                    throw new IllegalArgumentException("Levels Version 2 currently only supports SI");
```
Doesn't this also support just setting the unit? I swear I corrected some behavior for Ruth to get %, maybe it's a documentation thing though, the levels operate just like the timeseries as the units can be anything not just a length.

rma-rripken replied:
The getAll is querying against a AV_LOCATION_LEVEL and feeding the rows into the older JDom* classes. 
The Dao code for that method is clearly based on:
https://bitbucket.hecdev.net/projects/CWMS/repos/hec-cwms-data-access/browse/hec-db-jdbc/src/main/java/wcds/dbi/oracle/cwms/CwmsLevelJdbcDao.java

It looks to me like your changes were for calls from getOne for a single Level.  But maybe there are other pl/sql procs we should call instead of hitting the view.

MikeNeilson replied:
Well, that structure of this is fine, so we'll go for it for now. But someone is likely going to complain about not being able to set their preferred units.

rma-rripken committed:
checkstyle fixes. indentation, NotNull and unused variable.

rma-rripken committed:
checkstyle fixes. indentation, NotNull and unused variable.

adamkorynta committed:
finished rebasing changes for lazy loading CLOB/BLOB via urls

adamkorynta committed:
add text/binary value controllers for single large value download

adamkorynta committed:
add UML diagrams for large clob/blob retrieval

adamkorynta committed:
fix swagger UI not displaying non-CRUD endpoints

adding handlers as lambda expressions does not work as the generated anonymous inner class itself isn't annotated

MikeNeilson committed:
Use correct user for test.

dreina-gei committed:
Added else statement to error check to ensure all errors are thrown. Changed throw cause to throw e

rma-bryson committed:
Updated store ts text call in storeRow to use DATE_TABLE_TYPE

rma-bryson committed:
Fixed deserialization of ZonedDateTime by removing Json formatter pattern annotation. Added versioned date to unit test.

rma-bryson committed:
Updated versionDate field to an Instant in BinaryTimeSeries to be consistent. Removed overloaded "long" dateTime methods in BinaryTimeSeriesRow Builder. This resolves parsing issue with jackson attempting to parse an Instant string into a long.

rma-rripken committed:
Removing AtomicRef and using connectionResult instead.  Reformatted

rma-rripken committed:
Merge branch 'develop' into bugfix/level_duplicates

rma-rripken committed:
Merge branch 'develop' into bugfix/level_duplicates

rma-rripken committed:
Fixing merge conflict.
